[1.20.5] Implement specialized DataComponentType DeferredRegister

dhyces opened the pull request:
This PR aims to add a new specialized `DeferredRegister` for `DataComponentType`s. Due to data component types having a generic type, they require explicit type parameters when using regular registration methods. This PR fixes this using vanilla's own solution, a `UnaryOperator` with proper type inference. The `DeferredDataComponentType` also implements `DataComponentType` so that users may simply pass the deferred object to item property methods without calling `get`, which is simple convenience.

dhyces committed:
Implement specialized DataComponentType register

PepperCode1 commented:
I don't think implementing the registered object type (`DataComponentType`) on the holder is a good idea because the holders are not themselves registered so they cannot be used to look up the `ResourceLocation` of the registered object. The expectation when any code has access to an instance of any registered object type (such as `DataComponentType`) is that it has an associated `ResourceLocation` in the registry.

dhyces commented:
> I don't think implementing the registered object type (`DataComponentType`) on the holder is a good idea because the holders are not themselves registered so they cannot be used to look up the `ResourceLocation` of the registered object. The expectation when any code has access to an instance of any registered object type (such as `DataComponentType`) is that it has an associated `ResourceLocation` in the registry.

That's a good point, forgot about that when considering whether it was safe to do so. Removing it now.

EDIT: Actually, looking at it now, is there any reason to have a convenience type for this if it has no additional usable methods? Sure, it cuts down on generic verbosity, but otherwise it's just a regular `DeferredHolder`. Might as well just be a specialized registration method?

dhyces committed:
Remove DataComponentType interface implementation

Causes issues with reverse lookups to get associated ResourceLocation from the registry

ApexModder commented:
> EDIT: Actually, looking at it now, is there any reason to have a convenience type for this if it has no additional usable methods? Sure, it cuts down on generic verbosity, but otherwise it's just a regular DeferredHolder. Might as well just be a specialized registration method?

Yeah a specialized `DeferredHolder` type doesnt make much sense to me, the specialized register method is good though.
Could maybe also add a few other register methods which take a `Codec`/`StreamCodec` and wrap that into `UnaryOperator` method? 

<details>

<summary>Examples</summary>

```java
public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, UnaryOperator<DataComponentType.Builder<T>> builder) {
    return register(registrationName, () -> builder.apply(DataComponentType.builder()).build());
}

public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec, StreamCodec<? super RegistryFriendlyByteBuf, T> networkCodec) {
    return register(registrationName, builder -> builder.persistent(codec).networkSynchronized(networkCodec));
}

public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec, Codec<T> networkCodec) {
    return register(registrationName, codec, ByteBufCodecs.fromCodecWithRegistries(networkCodec));
}

public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec) {
    return register(registrationName, codec, codec);
}
```

</details>

To allow passing in the deferred type could patch in `Supplier` based component type get/set/has methods much like attachment types have, that way it would look up the correct registered object via the `Supplier` implementation, rather than using the deferred object itself.


Edit: Also looks like you have license header issues, run `gradlew applyAllFormatting` to apply neos spotless config, re-formats the code and applies licensing headers where need be, will get an error if theres any formatting violations.

and `gradlew unpackSourcePatches` to generate patch files
[_this is only needed if you modified vanilla code in `./projects/neoforge`, and best to be run before applying spotless config_]

Technici4n commented on src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java:
```diff
@@ -580,6 +595,32 @@ protected <I extends Item> DeferredItem<I> createHolder(ResourceKey<? extends Re
+         * 
+         * @param name    The name for this data component type. It will automatically have the {@linkplain #getNamespace() namespace} prefixed.
+         * @param builder The unary operator, which is passed a new builder for user operations, then builds it when registered.
+         * @return A {@link DeferredDataComponentType} that will track updates from the registry for this type.
```
there are no "registry updates" anymore

dhyces replied:
Docs for this were adapted from Items

Technici4n commented on src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java:
```diff
@@ -580,6 +595,32 @@ protected <I extends Item> DeferredItem<I> createHolder(ResourceKey<? extends Re
+         * @param builder The unary operator, which is passed a new builder for user operations, then builds it when registered.
+         * @return A {@link DeferredDataComponentType} that will track updates from the registry for this type.
+         */
+        public <D> DeferredDataComponentType<D> registerBuilder(String name, UnaryOperator<DataComponentType.Builder<D>> builder) {
```
the naming is not great - it's an actual component type that will be registered, not a builder

dhyces replied:
The object that the user will ever interact with in this case is a builder. Hence the shortened name `registerBuilder` as opposed to something like `registerWithBuilder`. Though, if I think of another name or another name is suggested, I can update it

Technici4n replied:
could be `registerComponent`

ApexModder replied:
why not simply call it `register`, its already inside of a type specific register `DeferredRegister.DataComponents`.

`Blocks` and `Items` dont call theirs `registerItem` or `registerBlock` so why should this be any different?
and the places where they do, are only named differently cause parameters clash or its for a sub-type (item block)

Technici4n replied:
![image](https://github.com/neoforged/NeoForge/assets/13494793/216a0e6e-d58f-4286-acc4-1e6a92c3090e)

They do call them like that... The ones that are called `register` are covariant overrides from `DeferredRegister`.

dhyces replied:
> why not simply call it `register`, its already inside of a type specific register `DeferredRegister.DataComponents`.
> 
> `Blocks` and `Items` dont call theirs `registerItem` or `registerBlock` so why should this be any different? and the places where they do, are only named differently cause parameters clash or its for a sub-type (item block)

In this case, it does clash, due to type erasure from `Function<RL, T>` and `UnaryFunction<T>` which extends `Function<T,T>`

Technici4n reviewed (changes requested):
I'm not sure that this feature is worth having as it makes the DeferredRegister API somewhat longer. It's probably fine given how widely used I'd expect data components to be.

dhyces commented:
> > EDIT: Actually, looking at it now, is there any reason to have a convenience type for this if it has no additional usable methods? Sure, it cuts down on generic verbosity, but otherwise it's just a regular DeferredHolder. Might as well just be a specialized registration method?
> 
> Yeah a specialized `DeferredHolder` type doesnt make much sense to me, the specialized register method is good though.
> Could maybe also add a few other register methods which take a `Codec`/`StreamCodec` and wrap that into `UnaryOperator` method? 
> 
> <details>
> 
> <summary>Examples</summary>
> 
> ```java
> public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, UnaryOperator<DataComponentType.Builder<T>> builder) {
>     return register(registrationName, () -> builder.apply(DataComponentType.builder()).build());
> }
> 
> public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec, StreamCodec<? super RegistryFriendlyByteBuf, T> networkCodec) {
>     return register(registrationName, builder -> builder.persistent(codec).networkSynchronized(networkCodec));
> }
> 
> public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec, Codec<T> networkCodec) {
>     return register(registrationName, codec, ByteBufCodecs.fromCodecWithRegistries(networkCodec));
> }
> 
> public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec) {
>     return register(registrationName, codec, codec);
> }
> ```
> 
> </details>
> 
> To allow passing in the deferred type could patch in `Supplier` based component type get/set/has methods much like attachment types have, that way it would look up the correct registered object via the `Supplier` implementation, rather than using the deferred object itself.
> 
> 
> Edit: Also looks like you have license header issues, run `gradlew applyAllFormatting` to apply neos spotless config, re-formats the code and applies licensing headers where need be, will get an error if theres any formatting violations.
> 
> and `gradlew unpackSourcePatches` to generate patch files
> [_this is only needed if you modified vanilla code in `./projects/neoforge`, and best to be run before applying spotless config_]

Thanks for a reminder about the licensing, I had forgotten that part of the workflow. I know how to apply patches and this PR was not in need of any patch updates.

dhyces commented:
> I'm not sure that this feature is worth having as it makes the DeferredRegister API somewhat longer. It's probably fine given how widely used I'd expect data components to be.

I'm inclined to agree, which is why I had mentioned a simple registration method instead. Only problem with that solution is that it bloats the regular generic registration methods with a specialized method which could only be used in registering data component types

dhyces committed:
Apply feedback and licensing

Minecraftschurli commented on src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java:
```diff
@@ -580,6 +595,32 @@ protected <I extends Item> DeferredItem<I> createHolder(ResourceKey<? extends Re
+         * @param builder The unary operator, which is passed a new builder for user operations, then builds it upon registration.
+         * @return A {@link DeferredDataComponentType} which reflects the data that will be registered.
+         */
+        public <D> DeferredDataComponentType<D> registerComponent(String name, UnaryOperator<DataComponentType.Builder<D>> builder) {
```
rename to `registerComponentType`

Minecraftschurli commented:
idk if the dedicated deferred holder is needed

dhyces committed:
Rename method and remove specialized holder
